src/msw/webview_edge.cpp

```diff
@@ -0,0 +1,866 @@
+{
+    m_impl->UpdateBounds();
+    event.Skip();
+}
+
+void wxWebViewEdge::LoadURL(const wxString& url)
+{
+    if (!m_impl->m_webView)
+    {
+        m_impl->m_pendingURL = url;
+        return;
+    }
+    HRESULT hr = m_impl->m_webView->Navigate(url.wc_str());
+    if (FAILED(hr))
+        wxLogApiError("WebView2::Navigate", hr);
+}
+
+void wxWebViewEdge::LoadHistoryItem(wxSharedPtr<wxWebViewHistoryItem> item)
+{
+    int pos = -1;
+    for (unsigned int i = 0; i < m_impl->m_historyList.size(); i++)
+    {
+        //We compare the actual pointers to find the correct item
+        if (m_impl->m_historyList[i].get() == item.get())
+            pos = i;
+    }
+    wxASSERT_MSG(pos != static_cast<int>(m_impl->m_historyList.size()), "invalid history item");
+    m_impl->m_historyLoadingFromList = true;
+    LoadURL(item->GetUrl());
+    m_impl->m_historyPosition = pos;
+}
+
+wxVector<wxSharedPtr<wxWebViewHistoryItem> > wxWebViewEdge::GetBackwardHistory()
+{
+    wxVector<wxSharedPtr<wxWebViewHistoryItem> > backhist;
+    //As we don't have std::copy or an iterator constructor in the wxwidgets
+    //native vector we construct it by hand
+    for (int i = 0; i < m_impl->m_historyPosition; i++)
+    {
+        backhist.push_back(m_impl->m_historyList[i]);
+    }
+    return backhist;
+}
+
+wxVector<wxSharedPtr<wxWebViewHistoryItem> > wxWebViewEdge::GetForwardHistory()
+{
+    wxVector<wxSharedPtr<wxWebViewHistoryItem> > forwardhist;
+    //As we don't have std::copy or an iterator constructor in the wxwidgets
+    //native vector we construct it by hand
+    for (int i = m_impl->m_historyPosition + 1; i < static_cast<int>(m_impl->m_historyList.size()); i++)
+    {
+        forwardhist.push_back(m_impl->m_historyList[i]);
+    }
+    return forwardhist;
+}
+
+bool wxWebViewEdge::CanGoForward() const
+{
+    if (m_impl->m_historyEnabled)
+        return m_impl->m_historyPosition != static_cast<int>(m_impl->m_historyList.size()) - 1;
+    else
+        return false;
+}
+
+bool wxWebViewEdge::CanGoBack() const
+{
+    if (m_impl->m_historyEnabled)
+        return m_impl->m_historyPosition > 0;
+    else
+        return false;
+}
+
+void wxWebViewEdge::GoBack()
+{
+    LoadHistoryItem(m_impl->m_historyList[m_impl->m_historyPosition - 1]);
+}
+
+void wxWebViewEdge::GoForward()
+{
+    LoadHistoryItem(m_impl->m_historyList[m_impl->m_historyPosition + 1]);
+}
+
+void wxWebViewEdge::ClearHistory()
+{
+    m_impl->m_historyList.clear();
+    m_impl->m_historyPosition = -1;
+}
+
+void wxWebViewEdge::EnableHistory(bool enable)
+{
+    m_impl->m_historyEnabled = enable;
+    m_impl->m_historyList.clear();
+    m_impl->m_historyPosition = -1;
+}
+
+void wxWebViewEdge::Stop()
+{
+    if (m_impl->m_webView)
+        m_impl->m_webView->Stop();
+}
+
+void wxWebViewEdge::Reload(wxWebViewReloadFlags WXUNUSED(flags))
+{
+    if (m_impl->m_webView)
+        m_impl->m_webView->Reload();
+}
+
+wxString wxWebViewEdge::GetPageSource() const
+{
```

Comment:
Is there a length limit on a string returned by `RunScript()` or the underlying API? If there is not, perhaps the page source could be returned by running `document.documentElement.outerHTML`? Similarly, the page text could be obtained by `document.body.textContent`. But TBH, my knowledge of javascript is basically nil, so...

BTW, in the sample I have noticed that the IE backend's `GetPageSource()` does not include the document type. When displaying the page source with the embedded browser context menu, the document type is there. I am MSW-only so I cannot test how this behaves with other backends. If needed, perhaps the document type could be obtained by calling `XMLSerializer().serializeToString(document.doctype)` and prepending it to the outerHTML.

Reply:
Implemented with your suggestions (used `innerText` instead of `textContent`).
I don't think we have to care about doctype for now if it's not currently returned by the IE backend.